pkg/managed/clusterroles_test.go

```diff
@@ -0,0 +1,281 @@
+	err = CleanupOrphanedClusterRoles(modelBindingPair, clusterConnections)
+	if err != nil {
+		t.Fatalf("got an error from CleanupOrphanedClusterRoles: %v", err)
+	}
+
+	// verify that the orphaned cluster role no longer exists
+	_, err = clusterConnection.KubeClient.RbacV1().ClusterRoles().Get(context.TODO(), "orphaned", metav1.GetOptions{})
+	if err == nil {
+		t.Fatal("expected cluster role to be cleaned up")
+	}
+}
+
+// TestCreateClusterRoles tests creation of cluster roles
+// GIVEN a cluster which does not have a cluster role with the service account name
+//  WHEN I call CreateClusterRoles
+//  THEN there should be a cluster role with the service account created
+//   AND that cluster role should have an expected set of policy rules
+func TestCreateClusterRoles(t *testing.T) {
+	assert := assert.New(t)
+
+	modelBindingPair := testutil.GetModelBindingPair()
+	clusterConnections := testutil.GetManagedClusterConnections()
+	clusterConnection := clusterConnections["cluster1"]
+
+	// the call to CreateClusterRoles should create a cluster role with the service account name
+	// the expectation is that it doesn't exist prior to the call
+	// verify that it doesn't exist
+	serviceAccountName := util.GetServiceAccountNameForSystem()
+	_, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().Get(context.TODO(), serviceAccountName, metav1.GetOptions{})
+	if err == nil {
+		t.Fatalf("did not expect cluster role %s to exist", serviceAccountName)
+	}
+
+	err = CreateClusterRoles(modelBindingPair, clusterConnections)
+	if err != nil {
+		t.Fatalf("got an error from CreateClusterRoles: %v", err)
+	}
+
+	// verify that a cluster role with the service account name has been created by the call to CreateClusterRoles
+	cr, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().Get(context.TODO(), serviceAccountName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("got an error trying to get a cluster role: %v", err)
+	}
+	// verify that the expected properties and policy rules are set on the new cluster role
+	assert.Equal(serviceAccountName, cr.Name, "the name of the cluster role should be %s", serviceAccountName)
+	assertGivenRoleHasExpectedRules(t, cr, policyRules)
+}
+
+// TestCreateClusterRolesVmiSystem tests creation of cluster roles when the binding name is 'system'
+// GIVEN a cluster which has no cluster roles
+//  WHEN I call CreateClusterRoles with a binding named 'system'
+//  THEN there should be an expected set of system cluster roles
+//   AND each system cluster role should have an expected set of policy rules
+func TestCreateClusterRolesVmiSystem(t *testing.T) {
+	modelBindingPair := testutil.GetModelBindingPair()
+	clusterConnections := testutil.GetManagedClusterConnections()
+	clusterConnection := clusterConnections["cluster1"]
+
+	// change the binding name to the VmiSystemBindingName so that CreateClusterRoles creates the system cluster roles
+	modelBindingPair.Binding.Name = constants.VmiSystemBindingName
```

Comment:
this feels like it is relying on a bug in V-O to make the test work :)

Reply:
do we really create the same clusterroles for the system vmi?

Reply:
What is the bug?  I'm just adjusting the starting test data (changing the binding name) to force a code path that I want to test.

When the binding name is VmiSystemBindingName then it creates what is termed SystemClusterRoles.  Again, I don't know exactly what the intent was here or if it is appropriate.  With all of these tests, unless I see something that is obviously wrong to me, I'm assuming that the current behavior is correct and trying to cover it with a reasonable set of assertions.

The code ...

```
	// Append Cluster roles from system monitoring
	if binding.Name == constants.VmiSystemBindingName {
		clusterRoles = append(clusterRoles, monitoring.GetSystemClusterRoles(clusterName)...)
		return clusterRoles
	}

```

```
// GetSystemClusterRoles gets all the cluster roles needed by Filebeats, Journalbeats, and NodeExporters in
// all the clusters.
func GetSystemClusterRoles(managedClusterName string) []*rbacv1.ClusterRole {
...
```

Reply:
yep, again - no need for you to do anything here.  it is the existing code that seems to have a problem, not the test you are adding.